src/peft/tuners/adamix.py

```diff
@@ -0,0 +1,416 @@
+        for name, module in self.model.named_modules():
+            if isinstance(module, ExpertSoup):
+                two_views.append(module.two_views)
+
+        if len(two_views) == 0:
+            raise ValueError("Returned views are empty")
+
+        return two_views
+
+
+def mark_only_adamix_as_trainable(model: nn.Module) -> None:
+    for n, p in model.named_parameters():
+        if "adamix" not in n:
+            p.requires_grad = False
+
+
+# # Below code is based on https://github.com/microsoft/AdaMix
+
+
+# #  ------------------------------------------------------------------------------------------
+# #  Copyright (c) Microsoft Corporation. All rights reserved.
+# #  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+# #  ------------------------------------------------------------------------------------------
+
+
+class MixtureSoup(nn.Module):
+    def __init__(self, expert, adapter_name, inference_mode=False, num_local_experts=1):
+        super().__init__()
+        self.inference_mode = inference_mode
+        self.dict_keys = ["_".join([adapter_name, str(i)]) for i in range(num_local_experts)]
+        self.expert_mixture = nn.ModuleDict({})
+        for i in self.dict_keys:
+            self.expert_mixture[i] = copy.deepcopy(expert)
+        self.num_local_experts = num_local_experts
+
+    def get_expert_by_idx(self, idx):
+        return self.expert_mixture[self.dict_keys[idx]]
+
+    def expert_soup_forward(self, x):
+        output = F.linear(x, self.parameter_dict["weight"], self.parameter_dict["bias"])
+        return output
+
+    def expert_soup(self):
+        self.parameter_dict = {"weight": 0, "bias": 0}
+        for idx in range(self.num_local_experts):
+            single_expert = self.expert_mixture[idx]
+            for s_name, s_param in single_expert.named_parameters():
+                if "weight" in s_name:
+                    p_name = "weight"
+                else:
+                    p_name = "bias"
+                self.parameter_dict[p_name] = self.parameter_dict[p_name] + s_param / self.num_local_experts
+
+    def forward(self, x: torch.Tensor):
+        expert_output = None
+
+        if not self.inference_mode:
+            expert_idx = torch.randint(low=0, high=self.num_local_experts, size=(1,)).item()  # selected expert
+            expert_output = self.get_expert_by_idx(expert_idx)(x)
+
+        else:
+            self.expert_soup()
+            expert_output = self.expert_soup_forward(x)
+
+        return expert_output
+
+
+class ExpertSoup(nn.Module):
+    def __init__(
+        self,
+        hidden_dim,
+        adapter_name,
+        adapter_dim,
+        num_expert=4,
+        sharing_down=False,
+        sharing_up=True,
+        return_two_views=False,
+        inference_mode=False,
+        dtype=torch.float16,
+    ):
+        super().__init__()
+
+        self.disable_adapters = False
+        self.return_two_views = return_two_views
```

Comment:
I wonder: Instead of having the user set `return_two_views` as a config parameter, could this decision be made based on whether the model is in training mode or eval mode? I think we only ever need it the two views during training, right?

Reply:
Correct. In eval mode, even if this flag is set to True, only one value is returned. The reason I had this flag is in case for some reason we only want one view during training